packages/driver/src/cypress/runner.ts

```diff
@@ -492,7 +492,57 @@ const hasOnly = (suite) => {
   )
 }
 
-const normalizeAll = (suite, initialTests = {}, setTestsById, setTests, getRunnableId, getHookId, getOnlyTestId, getOnlySuiteId, createEmptyOnlyTest) => {
+const pruneSuite = (emptySuite) => {
+  emptySuite.cleanReferences()
+
+  if (emptySuite.parent) {
+    const parentSuites = emptySuite.parent.suites.filter((suite) => suite !== emptySuite)
+
+    emptySuite.parent.suites = parentSuites
+    emptySuite.parent._onlySuites = emptySuite.parent._onlySuites.filter((suite) => parentSuites.includes(suite))
+    emptySuite.parent = null
+  }
+}
+
+const pruneEmptySuites = (rootSuite, incrementFoundTestsBy) => {
+  for (const suite of [...rootSuite.suites]) {
+    pruneEmptySuites(suite, incrementFoundTestsBy)
+  }
+
+  if (!rootSuite.suites.length && !rootSuite.tests.length) {
+    pruneSuite(rootSuite)
+  }
+
+  if (rootSuite.tests.length && Cypress.testFilter) {
+    incrementFoundTestsBy(rootSuite.tests.length)
+
+    const tests: any[] = []
+    const onlyTests: any[] = []
+
+    for (const test of rootSuite.tests) {
+      if (Cypress.testFilter.includes(test.fullTitle())) {
+        tests.push(test)
+
+        if (rootSuite._onlyTests.includes(test)) {
+          onlyTests.push(test)
+        }
+      } else {
+        delete test.fn
+      }
+    }
+
+    rootSuite.tests = tests
+    rootSuite._onlyTests = onlyTests
+
+    if (!rootSuite.tests.length && !rootSuite.suites.length) {
+      pruneSuite(rootSuite)
+    }
+  }
+}
+
+const normalizeAll = (suite, initialTests = {}, setTestsById, setTests, getRunnableId, getHookId, getOnlyTestId, getOnlySuiteId, createEmptyOnlyTest, incrementFoundTestsBy) => {
+  pruneEmptySuites(suite, incrementFoundTestsBy)
```

Comment:
To confirm, we will only prune in open mode correct?

Reply:
Can we add some unit tests around this behavior? https://github.com/cypress-io/cypress/blob/3925ae062a9b9dca4cc9d61391ce5f6ff0525a8b/packages/driver/cypress/e2e/cypress/runner.cy.js

Reply:
Not sure there is a good place to test this outside of the `cy-in-cy` test since the testFilter needs to be set before the spec file is executed.